ModifiedControlVariableCheck: NullPointerException on compact source files (JEP 512)

Parent Issue #19971

I have read check documentation: https://checkstyle.org/checks/coding/modifiedcontrolvariable.html
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

```bash
vivek@Viveks-MacBook-Air checkstyle % cat > config.xml <<'EOF'
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
    "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
    "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
  <module name="TreeWalker">
    <module name="ModifiedControlVariable"/>
  </module>
</module>
EOF

vivek@Viveks-MacBook-Air checkstyle % cat > InputModifiedControlVariableCompact.java <<'EOF'
void main() {
    for (int i = 0; i < 10; i++) {
        i++;
    }
}
EOF

vivek@Viveks-MacBook-Air checkstyle % RUN_LOCALE="-Duser.language=en -Duser.country=US"
java $RUN_LOCALE -jar target/checkstyle-13.6.0-SNAPSHOT-all.jar -c config.xml InputModifiedControlVariableCompact.java

Starting audit...
com.puppycrawl.tools.checkstyle.api.CheckstyleException: Exception was thrown while processing InputModifiedControlVariableCompact.java
        at com.puppycrawl.tools.checkstyle.Checker.processFiles(Checker.java:313)
        at com.puppycrawl.tools.checkstyle.Checker.process(Checker.java:227)
        at com.puppycrawl.tools.checkstyle.Main.runCheckstyle(Main.java:429)
        at com.puppycrawl.tools.checkstyle.Main.runCli(Main.java:347)
        at com.puppycrawl.tools.checkstyle.Main.execute(Main.java:205)
        at com.puppycrawl.tools.checkstyle.Main.main(Main.java:129)
Caused by: java.lang.NullPointerException: Cannot invoke "java.util.Deque.contains(Object)" because "currentVariables" is null
        at com.puppycrawl.tools.checkstyle.checks.coding.ModifiedControlVariableCheck.checkIdent(ModifiedControlVariableCheck.java:259)
        at com.puppycrawl.tools.checkstyle.checks.coding.ModifiedControlVariableCheck.visitToken(ModifiedControlVariableCheck.java:187)
        at com.puppycrawl.tools.checkstyle.TreeWalker.notifyVisit(TreeWalker.java:384)
        at com.puppycrawl.tools.checkstyle.TreeWalker.processIter(TreeWalker.java:457)
        at com.puppycrawl.tools.checkstyle.TreeWalker.walk(TreeWalker.java:322)
        at com.puppycrawl.tools.checkstyle.TreeWalker.processFiltered(TreeWalker.java:202)
        at com.puppycrawl.tools.checkstyle.api.AbstractFileSetCheck.process(AbstractFileSetCheck.java:101)
        at com.puppycrawl.tools.checkstyle.Checker.processFile(Checker.java:341)
        at com.puppycrawl.tools.checkstyle.Checker.processFiles(Checker.java:300)
        ... 5 more
Checkstyle ends with 1 errors.
```

---

**Describe what you expect in detail.**

ModifiedControlVariableCheck must not crash on a compact source file. A `for` loop that modifies its own loop variable inside a top-level method should be flagged as `Control variable 'i' is modified.`, the same as the equivalent loop inside an ordinary class.

Root cause: the check tracks loop variables on a `Deque<Deque<String>>` that is pushed in `enterBlock()` on every `OBJBLOCK` visit. Top-level methods in a compact source file have no `OBJBLOCK` ancestor (they are children of `COMPACT_COMPILATION_UNIT`), so the stack stays empty. When the assignment-style token is visited, `variableStack.peek()` returns null and `checkIdent` throws on `currentVariables.contains(...)` at `ModifiedControlVariableCheck.java:259`.

**Same code inside an ordinary class:**

```bash
vivek@Viveks-MacBook-Air checkstyle % cat > InputModifiedControlVariableOrdinary.java <<'EOF'
public class InputModifiedControlVariableOrdinary {
    public void main() {
        for (int i = 0; i < 10; i++) {
            i++;
        }
    }
}
EOF

vivek@Viveks-MacBook-Air checkstyle % java $RUN_LOCALE -jar target/checkstyle-13.6.0-SNAPSHOT-all.jar -c config.xml InputModifiedControlVariableOrdinary.java

Starting audit...
[ERROR] InputModifiedControlVariableOrdinary.java:4:14: Control variable 'i' is modified. [ModifiedControlVariable]
Audit done.
Checkstyle ends with 1 errors.
```

--- a/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/ModifiedControlVariableCheck.java
+++ b/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/ModifiedControlVariableCheck.java
@@ -126,6 +126,7 @@ public final class ModifiedControlVariableCheck extends AbstractCheck {
     public int[] getRequiredTokens() {
         return new int[] {
             TokenTypes.OBJBLOCK,
+            TokenTypes.COMPACT_COMPILATION_UNIT,
             TokenTypes.LITERAL_FOR,
             TokenTypes.FOR_ITERATOR,
             TokenTypes.FOR_EACH_CLAUSE,
@@ -162,7 +163,8 @@ public final class ModifiedControlVariableCheck extends AbstractCheck {
     @Override
     public void visitToken(DetailAST ast) {
         switch (ast.getType()) {
-            case TokenTypes.OBJBLOCK -> enterBlock();
+            case TokenTypes.OBJBLOCK,
+                 TokenTypes.COMPACT_COMPILATION_UNIT -> enterBlock();
             case TokenTypes.LITERAL_FOR,
                  TokenTypes.FOR_ITERATOR,
                  TokenTypes.FOR_EACH_CLAUSE -> {
@@ -200,7 +202,8 @@ public final class ModifiedControlVariableCheck extends AbstractCheck {
                 }
             }
             case TokenTypes.LITERAL_FOR -> leaveForDef(ast);
-            case TokenTypes.OBJBLOCK -> exitBlock();
+            case TokenTypes.OBJBLOCK,
+                 TokenTypes.COMPACT_COMPILATION_UNIT -> exitBlock();
             case TokenTypes.ASSIGN,
                  TokenTypes.PLUS_ASSIGN,
                  TokenTypes.MINUS_ASSIGN,

--- a/src/test/java/com/puppycrawl/tools/checkstyle/checks/coding/ModifiedControlVariableCheckTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/checks/coding/ModifiedControlVariableCheckTest.java
@@ -172,6 +172,32 @@ public class ModifiedControlVariableCheckTest
                 expected);
     }
 
+    @Test
+    public void testCompactSourceFile() throws Exception {
+        final String[] expected = {
+            "11:10: " + getCheckMessage(MSG_KEY, "i"),
+            "19:11: " + getCheckMessage(MSG_KEY, "k"),
+            "24:14: " + getCheckMessage(MSG_KEY, "a"),
+            "31:10: " + getCheckMessage(MSG_KEY, "i"),
+            "36:14: " + getCheckMessage(MSG_KEY, "item"),
+            "43:14: " + getCheckMessage(MSG_KEY, "i"),
+        };
+        verifyWithInlineConfigParser(
+                getNonCompilablePath("InputModifiedControlVariableCompactSourceFile.java"),
+                expected);
+    }
+
+    @Test
+    public void testCompactSourceFileSkipEnhancedForLoop() throws Exception {
+        final String[] expected = {
+            "16:10: " + getCheckMessage(MSG_KEY, "i"),
+        };
+        verifyWithInlineConfigParser(
+                getNonCompilablePath(
+                    "InputModifiedControlVariableCompactSourceFileSkipEnhancedForLoop.java"),
+                expected);
+    }
+
     /**
      * We cannot reproduce situation when visitToken is called and leaveToken is not.
      * So, we have to use reflection to be sure that even in such situation

--- a/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/coding/modifiedcontrolvariable/InputModifiedControlVariableCompactSourceFile.java
+++ b/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/coding/modifiedcontrolvariable/InputModifiedControlVariableCompactSourceFile.java
@@ -1,0 +1,46 @@
+/*
+ModifiedControlVariable
+skipEnhancedForLoopVariable = (default)false
+
+
+*/
+// non-compiled with javac: Compilable with Java25
+
+void main() {
+    for (int i = 0; i < 10; i++) {
+        i++; // violation, 'Control variable 'i' is modified'
+    }
+
+    for (int j = 0; j < 10; j++) {
+        System.out.println(j);
+    }
+
+    for (int k = 0; k < 10; k++) {
+        k += 2; // violation, 'Control variable 'k' is modified'
+    }
+
+    for (int a = 0; a < 10; a++) {
+        for (int b = 0; b < 10; b++) {
+            a++; // violation, 'Control variable 'a' is modified'
+        }
+    }
+}
+
+void second() {
+    for (int i = 0; i < 5; i++) {
+        i--; // violation, 'Control variable 'i' is modified'
+    }
+
+    String[] items = {"x", "y"};
+    for (String item : items) {
+        item = "z"; // violation, 'Control variable 'item' is modified'
+    }
+}
+
+class Inner {
+    void m() {
+        for (int i = 0; i < 3; i++) {
+            i++; // violation, 'Control variable 'i' is modified'
+        }
+    }
+}

--- a/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/coding/modifiedcontrolvariable/InputModifiedControlVariableCompactSourceFileSkipEnhancedForLoop.java
+++ b/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/coding/modifiedcontrolvariable/InputModifiedControlVariableCompactSourceFileSkipEnhancedForLoop.java
@@ -1,0 +1,18 @@
+/*
+ModifiedControlVariable
+skipEnhancedForLoopVariable = true
+
+
+*/
+// non-compiled with javac: Compilable with Java25
+
+void main() {
+    String[] items = {"x", "y"};
+    for (String item : items) {
+        item = "z"; // ok, enhanced for-loop variable is skipped
+    }
+
+    for (int i = 0; i < 5; i++) {
+        i++; // violation, 'Control variable 'i' is modified'
+    }
+}

